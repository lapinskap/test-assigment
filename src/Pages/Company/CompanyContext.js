import React, {
  useContext, useEffect, useState, useCallback,
} from 'react';
import PropTypes from 'prop-types';
import { COMPANY_MANAGEMENT_SERVICE, restApiRequest } from '../../utils/Api';
import { mockData } from './CompanyList';
import ApiForbiddenError from '../../utils/Api/ApiForbiddenError';
import ForbiddenErrorPage from '../../utils/ErrorPages/403';

const CompanyContext = React.createContext({
  data: {},
  refresh: () => null,
});

export const CompanyContextWrapper = ({ companyId, children, showError }) => {
  const [data, setData] = useState({ id: companyId });
  const [hasAccess, setHasAccess] = useState(true);
  const fetchCompanyData = useCallback(() => {
    if (companyId) {
      restApiRequest(
        COMPANY_MANAGEMENT_SERVICE,
        `/companies/${companyId}`,
        'GET',
        {},
        mockData.find(({ id }) => id === companyId),
      ).then((res) => {
        if (res) {
          setData(res);
        }
      }).catch((e) => {
        if (e instanceof ApiForbiddenError) {
          setHasAccess(false);
        } else {
          console.error(e);
        }
      });
    }
  }, [companyId]);
  useEffect(() => {
    fetchCompanyData();
  }, [companyId, fetchCompanyData]);
  let content = children;
  if (!hasAccess) {
    content = showError ? <ForbiddenErrorPage /> : null;
  }
  return (
    <CompanyContext.Provider value={{
      data,
      refresh: fetchCompanyData,
    }}
    >
      {content}
    </CompanyContext.Provider>
  );
};

CompanyContextWrapper.propTypes = {
  companyId: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
  showError: PropTypes.bool,
};

CompanyContextWrapper.defaultProps = {
  showError: true,
};

export const useCompanyName = () => {
  const { data = {} } = useContext(CompanyContext);
  const { fullName: name = '' } = data;
  return name;
};

export const useCompanyId = () => {
  const { data = {} } = useContext(CompanyContext);
  const { id = '' } = data;
  return id;
};

export const useCompanyHasFunctionality = (code) => {
  const { data = {} } = useContext(CompanyContext);
  const { functionalities = [] } = data;
  return functionalities.includes(code);
};

export const useCompanyValue = (fieldCode, defaultValue = null) => {
  const { data = {} } = useContext(CompanyContext);
  return data[fieldCode] || defaultValue;
};

export const useCompanyData = () => {
  const { data = {} } = useContext(CompanyContext);
  return data;
};

export default CompanyContext;
