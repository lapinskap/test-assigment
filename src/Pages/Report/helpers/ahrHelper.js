export const getAhrUrl = (url, isAhr) => {
  let result = isAhr ? '/ahr' : '';
  result += url;

  return result;
};

export const asd = () => {};
