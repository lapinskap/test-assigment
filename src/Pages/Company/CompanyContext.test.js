/* eslint-disable react/prop-types */
import React from 'react';
import { renderHook } from '@testing-library/react-hooks';
import { create } from 'react-test-renderer';
import CompanyContext, {
  CompanyContextWrapper, useCompanyName, useCompanyHasFunctionality, useCompanyValue,
} from './CompanyContext';
import { wait, waitToLoadMocks } from '../../tests/helpers';

describe('CompanyContext', () => {
  test('CompanyContext renders correctly', async () => {
    const wrapper = create((
      <CompanyContextWrapper companyId="a43275e4-eeb2-11ea-adc1-0242ac1200021" showError>
        <div className="text-wrap">
          Zarządzanie firmą
          <CompanyContext.Consumer>
            {({ data }) => data.fullName }
          </CompanyContext.Consumer>
        </div>
      </CompanyContextWrapper>
    ));
    await wait(150);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});

describe('useCompanyName', () => {
  it('returns company name', () => {
    const wrapper = ({ children }) => (
      <CompanyContext.Provider value={{
        data: {
          fullName: 'test name',
        },
        refresh: () => null,
      }}
      >
        {children}
      </CompanyContext.Provider>
    );
    const { result } = renderHook(() => useCompanyName(), { wrapper });
    expect(result.current).toBe('test name');
  });
});

describe('useCompanyValue', () => {
  it('returns company value', async () => {
    const wrapper = ({ children }) => (
      <CompanyContext.Provider value={{
        data: {
          name: 'test',
        },
        refresh: () => null,
      }}
      >
        {children}
      </CompanyContext.Provider>
    );
    const { result } = renderHook(() => useCompanyValue('name', ''), { wrapper });
    await waitToLoadMocks();
    expect(result.current).toBe('test');
  });
  it('returns default value', () => {
    const wrapper = ({ children }) => (
      <CompanyContext.Provider value={{
        data: {
        },
        refresh: () => null,
      }}
      >
        {children}
      </CompanyContext.Provider>
    );
    const { result } = renderHook(() => useCompanyValue('name', ''), { wrapper });
    expect(result.current).toBe('');
  });
});

describe('useCompanyHasFunctionality', () => {
  it('returns true', () => {
    const wrapper = ({ children }) => (
      <CompanyContext.Provider value={{
        data: {
          name: '',
          functionalities: ['EMPLOYEE_PESEL'],
        },
        refresh: () => null,
      }}
      >
        {children}
      </CompanyContext.Provider>
    );
    const { result } = renderHook(() => useCompanyHasFunctionality('EMPLOYEE_PESEL'), { wrapper });
    expect(result.current).toBe(true);
  });
  it('returns false', () => {
    const wrapper = ({ children, data }) => (
      <CompanyContext.Provider value={{
        data,
      }}
      >
        {children}
      </CompanyContext.Provider>
    );
    const { result } = renderHook(() => useCompanyHasFunctionality('test'), { wrapper });
    expect(result.current).toBe(false);
  });
});
