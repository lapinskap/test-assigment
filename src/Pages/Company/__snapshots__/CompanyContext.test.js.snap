// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompanyContext CompanyContext renders correctly 1`] = `
<div
  className="text-wrap"
>
  Zarządzanie firmą
  10clouds
</div>
`;
